### ISSUE TITLE
Fast rollout features

### ISSUE THREAD

**Issue by wrongu** [Rochester-NRT/AlphaGo#48]
Need to implement features described in Extended Data Table 4

- [ ]  Response
- [ ]  Save atari
- [ ] Neighbor
- [ ] Nakade
- [ ] Response pattern
- [ ] Non-response pattern
- [ ] Last move distance

Note that 'self-atari' in this network can be the 'self atari' feature for the deep policy network with `maximum=1`

**Comment by LetterRip** [comment]
Did some thinking on this last night.  For the 'last move distance', my understanding it is the Manhattan Distance from the last two moves.  It supposedly gives 34 possible patterns.   Is it the sum of distances?  ie MHDist(p1, p2) + MHDist(p2, p3) or MHDist(p1, p3)+MHDist(p2, p3) or a tuple?  Or something else?  assuming it is the sum of distances - for the first dist method I get 53 possibilities and for the second I get 70 possibilities, both of which are quite a bit more than 34.  So need some clarification there.

For the patterns - are there pattern databases that we should be looking at?  I don't think the patterns can be copyrighted (they are 'merely functional')

**Comment by LetterRip** [comment]
It looks like they might be using the LGPLed libraries from Fuego, since the features it uses for decisions are basically the same list, here is the relevant code.

https://sourceforge.net/p/fuego/code/HEAD/tree/trunk/features/FeBasicFeatures.cpp

https://users.soe.ucsc.edu/~glin/docs/Fuego_Fall09Report.pdf

**Comment by LetterRip** [comment]
Here is the relevant code I've found so far...

Save Atari GoBoardUtil.h
Nakade	GoEyeUtil.cpp, .h
Non-response pattern -> GoPattern3x3.cpp
Response pattern -> GoPattern12Point.cpp

Self-atari -> GoBoardUtil.h
Last move distance -> FeBasicFeatures.cpp
Non-respnse pattern -> GoPattern12Point.cpp

**Comment by LetterRip** [comment]
So a question I have, that isn't clear from the paper - for the patterns - is this a massive vector that each pattern is a one hot?  or is it a single variable with a 1 if any of the patterns are present.

**Comment by wrongu** [comment]
Thanks for all the resources! Regarding your last question, we interpreted it as a massive 1-hot encoding. Notice that the 'response' feature says "Whether move matches __one or more__ response pattern features". So the features with >1 pattern must mean all of the possible patterns.. Actually building the full one-hot encoding for a 69338-dimensional vector and multiplying by a weight vector could be replaced by a quick lookup of the `i`th softmax weight if it turns out to be slow.

**Comment by LetterRip** [comment]
I emailed Aja for clarification, will let you know what they say.  I figured it is probably one hot encoding as well but doesn't hurt to ask, since I already needed clarification on their manhattan distance algorithm.

**Comment by LetterRip** [comment]
This is the function I have so far for the modified manhattan distance using information from Fuego.  Note that addition of max(dx, dy) and truncation which makes it not a true manhattan distance.

def truncatedManhattanDistance(p1, p2):
    dx = abs(p1.x - p2.x)
    dy = abs(p1.y - p2.y)
    d = min(dx + dy + max(dx, dy), 17)
    return d

def summedTruncatedManhattanDistancs(p1, p2, p3):
    return truncatedManhattanDistance(p1, p3)+
        truncatedManhattanDistance(p2, p3)

**Comment by LetterRip** [comment]
Looks like the 3x3 pattern code (and some other stuff as well) will be easier to adapt from here

https://github.com/pasky/michi/blob/master/michi.py

and for speed here is a c reimplementation of the same...

https://github.com/db3108/michi-c

Both are MIT licensed, which is good.

**Comment by LetterRip** [comment]
Ah the distance I thought they were using for Manhattan distance, I misunderstood it.  It is actually 'for gridcular' coordinates for the diamond pattern.

https://groups.google.com/forum/#!topic/computer-go-archive/2pQtpqyxSq8

**Comment by LetterRip** [comment]
Making good progress, will probably have a preliminary commit ready in the next day or so that has all or almost all of the features ready (though I'm not getting the same pattern counts...)

**Comment by wrongu** [comment]
I tried enumerating _all possible_ 3x3 patterns where the middle space is empty, taking into account the number of liberties (a detail in the paper methods), pruning out all symmetries, etc.. and came up with a number around 45000 total. Yet in the paper they claim to use 69338 3x3 patterns. Either they use some duplicate/impossible patterns, or I have a bug in my code. The bug seems likely at this point..

**Comment by wrongu** [comment]
My sanity-check counts were way off... I just re-did them and found 486578 possible 3x3 patterns where

* stone color and liberties (1,2,>=3) jointly make a position unique (from the paper's methods)
* the middle is empty
* all symmetric positions are removed
* connected stones have the same (1,2, or >=3) liberties
* each group has at least as many liberties as adjacent empty spaces visible in the 3x3 area
* I ignored patterns with the edge of the board, which would only increase the number

So 69338 possible variations within the patterns used in Fuego may be more reasonable than I had thought. How did you check these pattern counts?

**Comment by LetterRip** [comment]
I'm currently only matching the patterns that are used in other Go systems.  See the michi code, which I've based it on...

https://github.com/pasky/michi/blob/master/michi.py

plus see my pull request (this filters out the impossible patterns)
https://github.com/pasky/michi/pull/11

Then I'll add in the liberty counts as well coloration information...

Really I think we should reduce the pattern count dramatically since the variants collapse to 14 or so common situations (double it for the color of the pattern...) then a multiplier again for number of liberties.

**Comment by wrongu** [comment]
@LetterRip are you still working on this? If not, could you make a PR with the features you implemented?

**Comment by LetterRip** [comment]
@wrongu sorry for the delayed response.  No I'm not working on this.  I kept running into the problem that my features would be orders of magnitude different from those in their paper.  So I suspect I simply don't understand go well enough to do the implementations.

**Comment by MaMiFreak** [comment]
AG uses two fast policies: fast rollout policy and tree policy. 
-- rollout policy
- [ ] Response:   1 -> match response pattern
- [ ] Save Atari: 1 -> move saves stones (i assume no ladder check?)
- [ ] Neighbor:   8 -> 8-connected 
- [ ] Nakade:     8192 -> match nakade pattern at captured stone (why captured only?)
- [ ] Response pattern: 32207 -> 12-point diamond pattern near previous move
- [ ] Non-response pattern: 69338 -> 3x3 pattern around move 
-- tree policy
- [ ] Self-atari           
- [ ] Last move distance   -> manhattan distance with max 34
- [ ] Non-response pattern -> 12-point diamond pattern around move

Explanation Extended Data Table 4 in paper:
"Features used by the rollout policy (first set) and tree policy (first and second set). Patterns are based on stone colour (black/white/empty) and liberties (1, 2, ≥3) at each intersection of the pattern."

AG only uses `tree policy` to create a placeholder next move prediction for a new MTCS tree node until a `SL policy`  analysis is preformed. Right now there is no `tree policy` included in MTCS framework so i only focused on the rollout policy features.

I gathered 3x3 and 12-point diamond shape patterns from ~51000 pro sgf files, [results can be found here](https://github.com/MaMiFreak/RocAlphaGo/tree/debug-rollout/rollout_data). 

To me the most promising seems removing all rotational and colour symmetries taking the top ~4500 patterns (patterns used >~90) restoring all rotational and colour symmetries we have a ~69000 pattern count but only need a ~4500 one-hot vector

same goes for the 12D patterns  removing all rotational and colour symmetries taking the top ~2000 patterns (patterns used >~150) restoring all rotational and colour symmetries we have a ~32000pattern count but only need a ~2000one-hot vector

- as discussed before [Rémi Coulom](http://www.grappa.univ-lille3.fr/~coulom/ChineseGo/MMGoPatterns.pdf) suggest just using the patterns most frequently used has its drawbacks and suggests to calculate an elo ranking for each pattern on how successful it is.

- an idea i was wondering about is to train a small network using all patterns and use the ones with the highest bias.

next to that i had a look at the nakade patterns but i am in doubt about their description: `match nakade pattern at captured stone` what do they mean with  "at captured stone"? it seems unlikely to me but are they only used after a stone capture?

i created a small collection of dead shapes (in nakade folder) and try to come up with an useful way to create a hash for them but so far i have not found a good way to do so. suggestions are welcome! 

@wrongu & @prescod what kind of lookup would be most useful/fastest? also what kind of hash would you suggest?

@oriskunk you are working on using pachi rollout patterns, do you have to gather patterns to use with it? i only had a look at fuego and gnugo.

**Comment by wrongu** [comment]
It's been a while since I've looked at this. What's the count of 3x3 and 12D patterns used within fuego/pachi themselves?

Not sure what a fast way to do this would be. In lieu of a solution, here are some random thoughts:

* A simple implementation would be to store all 69000/32000 lookup values in a `dict`. Given a board, loop over `(x,y)` and lookup the dict value for each one. __Pro:__ easy to implement. No transformations required. __Con:__ wasted space, CPU-bound computation.
* To overcome wasted space, as you suggested, decide on a 'representative' instance of each pattern. Do this by _ordering_ all transforms of a position and taking the least (lexicographically or by hash, e.g.) of each as the representative. __Pro:__ less wasteful of space. __Con:__ extra lookup time for each pattern.
* Note that CPU-boundness may not be too bad. Imagine that we put these lookup dicts onto the GPU.. you'd still have to compute liberties and send the tensor to the GPU in relatively small batches. I'd expect that the communication overhead is not worth the parallelism.
* I see no real reason to expect a hash table/dict to be any slower than, say, a list indexed by the lexicographic rank order of patterns. In general that might make sense if patterns are accessed in a predictable order, since you could better use your CPU caches. But I wouldn't expect a pattern to the order in which 3x3-patterns are seen, e.g...
* Having spelled this all out, I am now leaning towards the simple implementation, using a `dict` for each LUT...

**Comment by MaMiFreak** [comment]
fuego: only uses handcrafted 3x3 patterns and no 12D as far as i was able to determine, only stone positions are used no liberty counts.

pachi: i am still figuring out how they implemented it. seems by default they are not used but pachi is able to generate 3x3 patterns. http://pachi.or.cz/pat/ has some libraries people can download (~102000patterns) but i do not know if rotational & colour symmetries are included.. 12D patterns (and several others as well) are available but i have found no libraries for them
@oriskunk maybe you know more about pachi pattern implementation?

i am also in favor of the simple implementation, i think the additional space outweighs the computation time of any rotation before lookup and RL is CPU bound in AG as well.

Thanks! your summary helped with some other things as well. going to redesign the 12D to be able to use direct (x,y)liberty count lookup as wel.

**Comment by MaMiFreak** [comment]
is there a specific reason to use -1/0/1 for white/empty/black? i would like to change them to 0/1/2 ( 0 = empty ) it would be rather useful when creating pattern values.

i made an initial commit with model and my idea on (preprocessing structure)[https://github.com/MaMiFreak/RocAlphaGo/blob/debug-rollout/AlphaGo/preprocessing/preprocessing.py], my thought is to hold pattern dict in Gamestate and to use preprocessing as it is.

**Comment by wrongu** [comment]
-1/0/1 is convenient for the value net among other things. I doubt `+1` will have a big impact on lookup

**Comment by MaMiFreak** [comment]
the new value net is relative to active player so it does no longer use it but even so a single position evaluation with value net would encounter this -1 a maximum of 361x per position.
a single fast rollout would have to do this 5500 times on an empty board for the 3x3 & 12d patterns, nakade patterns not included and then for 300/400 moves. 

right now the impact does not matter but at some point these small improvements will become important=)
i will use the `+1` lookup

**Comment by MaMiFreak** [comment]
performed a successful first test with 3x3 patterns. managed to archief 10.5%acc but with a very limited dataset. 
- 12d patterns are needed before i can test any game performance, right now it is unable to create any living group at all.. 
- peformance is poor, have to check wheter it is my quick&dirty feature implementation or the 4500 one-hot tensors. (right now rollout move generation takes longer than SL move generation on cpu)
- Nakade patterns are still a problem, i have not found a way to hash them in any way or to get 8000 of them...
